File: Jan2021/05Jan/RemoveDuplicatesFromSortedListII.cpp
/*
Problem name: 82. Remove Duplicates from Sorted List II
Problem link: https://leetcode.com/problems/remove-duplicates-from-sorted-list-ii/
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        ListNode* temp=new ListNode(-1);
        temp->next=head;
        ListNode* prev=temp;
        ListNode* curr=head;

        while(curr){
            while(curr->next && curr->val == curr->next->val)
                curr=curr->next;
            
            if(prev->next==curr){
                prev=prev->next;
                curr=curr->next;
            }
            else{
                prev->next=curr->next;
                curr=curr->next;
            }
        }
        return temp->next;
    }
};

File: Jan2021/05Jan/RemoveDuplicatesFromSortedListII.java
/*
Problem Name: 82. Remove Duplicates from Sorted List II
Problem Link: https://leetcode.com/problems/remove-duplicates-from-sorted-list-ii/
*/
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode deleteDuplicates(ListNode head) {
        if(head == null)
            return head;
        ListNode ancestor = new ListNode(-1, head), prev = ancestor;
        while(head != null){
            if(head.next != null && head.val == head.next.val){
                while(head.next != null && head.val == head.next.val)
                    head = head.next;
                prev.next = head.next;
            }
            else
                prev = prev.next;
            head = head.next;
        }
        return ancestor.next;
    }
}
